lib/logstash/config/config_ast.rb

```diff
@@ -211,13 +216,7 @@ def compile
         return "start_input(#{variable_name})"
       when "filter"
         return <<-CODE
-          events = events.flat_map do |event|
-            next [] if event.cancelled?
-
-            new_events = []
-            #{variable_name}.filter(event){|new_event| new_events << new_event}
-            event.cancelled? ? new_events : new_events.unshift(event)
-          end
+          events = #{variable_name}.multi_filter(events)
         CODE
```

Comment:
I already explained some of this in the issue comments: the events cancellation is now dealt with in the pipeline code. It is not for the shutdown but when I filter marks an event not to be sent to the output queue, think multiline: it will aggregate events and cancel them util the multiline boundary is reached. 

`flat_map` is not required anymore, previously each filter step was pushing an array in the `events` array but now only individual events are pushed so `flat_map` is not needed anymore.